artiq/gateware/rtio/phy/urukul_monitor.py

```diff
@@ -55,23 +55,38 @@ def update_probe_data(i):
 
         # 0 -> init, 1 -> start reading
         state = [Signal(1) for i in range(nchannels)]
```

Comment:
FSM()

Reply:
Let's not change it now since I can't find ways to fix it, for some reason it doesn't work:
```py
        for i in range(nchannels):
            fsm = ClockDomainsRenamer("rio_phy")((FSM(reset_state="IDLE")))
            fsm.act("IDLE", [
                If(self.selected(i + CS_DDS_CH0) & (self.current_address == SPI_DATA_ADDR), [
                    If(~self.current_data[31], [
                        If((AD9910_REG_PROFILE0 <= reg) & (reg <= AD9910_REG_PROFILE7), [
                            # assume it is write64, which means the last command in the SPI sequence (SPI_END)
                            # will be FTW
                            NextState("READ_NEXT")
                        ]),
                        If(reg == AD9910_REG_FTW, [
                            If((self.length == 24) & (self.flags & SPI_END), [
                                # write16
                                buffer[i][:16].eq(self.current_data[8:24]),
                            ]),
                            If(self.length == 8, [
                                # write32
                                NextState("READ_NEXT")
                            ])
                        ])
                    ]),
                ])
            ])

            fsm.act("READ_NEXT", [
                If(self.selected(i + CS_DDS_CH0) & (self.current_address == SPI_DATA_ADDR), [
                    If(self.flags & SPI_END, [
                        buffer[i][:32].eq(self.current_data),
                        NextState("IDLE")
                    ])
                ])
            ])

            self.submodules += ClockDomainsRenamer("rio_phy")(fsm)

```